framework/src/auxkernels/NodalPatchRecovery.C

```diff
@@ -0,0 +1,224 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NodalPatchRecovery.h"
+#include "SwapBackSentinel.h"
+#include "libmesh/utility.h"
+
+template <>
+InputParameters
+validParams<NodalPatchRecovery>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addParam<unsigned>("patch_polynomial_order", "polynomial order");
+
+  // TODO make this class work with relationship manager
+  // params.registerRelationshipManagers("ElementSideNeighborLayers");
+  // params.addPrivateParam<unsigned short>("element_side_neighbor_layers", 2);
+
+  return params;
+}
+
+NodalPatchRecovery::NodalPatchRecovery(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _order_provided(isParamValid("patch_polynomial_order")),
+    _order(_order_provided ? getParam<unsigned>("patch_polynomial_order") : (unsigned)_var.order()),
+    _dim(_mesh.dimension()),
+    _fe_problem(*parameters.get<FEProblemBase *>("_fe_problem_base")),
+    _multi_index(multiIndex(_dim, _order))
+{
+  // TODO remove the manual ghosting once relationship manager is working correctly
+  // no need to ghost if this aux is elemental
+  if (isNodal())
+  {
```

Comment:
Would it make sense to have a check to make sure that the user didn't specify a patch polynomial order lower than the variable order?
```
   if (isParamValid("patch_polynomial_order"))
    mooseWarning("Specified 'patch_polynomial_order' is lower than the AuxVariable's order");
```

Reply:
I think it is okay at least in some cases to specify a patch polynomial order higher than the variable order as we discussed earlier. I do have a check to make sure we have enough patch sample points to do the least squares fitting:

```
// before we go, check if we have enough sample points for solving the least square fitting
  if (_q_point.size() * elem_ids.size() < _multi_index.size())
    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
               "polynomial order or using a higher-order quadrature scheme.");
```